javagen/src/main/java/com/azure/autorest/template/StreamSerializationModelTemplate.java

```diff
@@ -101,71 +121,165 @@ protected void addModelConstructorParameter(ClientModelProperty property,
+        });
+
+        classBlock.publicStaticMethod(wrapperClassName + " fromXml(XmlReader xmlReader) throws XMLStreamException", readerMethod -> {
+            String readObject = (xmlNamespace != null)
+                ? "return xmlReader.readObject(\"" + xmlNamespace + "\", \"" + xmlRootElementName + "\", reader -> {"
+                : "return xmlReader.readObject(\"" + xmlRootElementName + "\", reader -> {";
+
+            readerMethod.line(readObject);
+            readerMethod.indent(() -> {
+                readerMethod.line(iterableType + " items = null;");
+                readerMethod.line();
+                readerMethod.line("while (reader.nextElement() != XmlToken.END_ELEMENT) {");
+                readerMethod.indent(() -> {
+                    // TODO (alzimmer): Support namespace validation.
+                    readerMethod.line("String elementName = reader.getElementName().getLocalPart();");
+                    readerMethod.line();
+                    readerMethod.ifBlock("\"" + xmlListElementName + "\".equals(elementName)", ifBlock -> {
+                        ifBlock.ifBlock("items == null", ifBlock2 -> ifBlock2.line("items = new ArrayList<>();"));
+                        ifBlock.line();
+
+                        // TODO (alzimmer): Insert XML object reading logic.
+                        ifBlock.line("items.add(" + getSimpleXmlDeserialization(elementType, elementType, "reader", null, null) + ");");
+                    }).elseBlock(elseBlock -> elseBlock.line("reader.nextElement();"));
+                });
+                readerMethod.line("}");
+
+                readerMethod.methodReturn("new " + wrapperClassName + "(items)");
+            });
+            readerMethod.line("});");
+        });
+    }
+
     @Override
     protected void addFieldAnnotations(ClientModelProperty property, JavaClass classBlock, JavaSettings settings) {
         // no-op as stream-style serialization doesn't add any field-level annotations.
     }
 
     @Override
     protected void writeStreamStyleSerialization(JavaClass classBlock, ClientModel model, JavaSettings settings) {
-        PropertiesManager propertiesManager = new PropertiesManager(model);
+        // Early out as strongly-typed headers do their own thing.
+        if (model.isStronglyTypedHeader()) {
+            return;
+        }
+
+        ClientModelPropertiesManager propertiesManager = new ClientModelPropertiesManager(model, settings);
 
-        writeToJson(classBlock, propertiesManager);
-        writeFromJson(classBlock, model, propertiesManager, settings);
+        if (settings.isGenerateXmlSerialization() && model.getXmlName() != null) {
+            writeToXml(classBlock, propertiesManager);
+            writeFromXml(classBlock, model, propertiesManager, settings);
+        } else {
+            writeToJson(classBlock, propertiesManager);
+            writeFromJson(classBlock, model, propertiesManager, settings);
+        }
```

Comment:
nit, given the size of this file, wondering that if JSON and XML is orthogonal, will it worthwhile to put them to separate classes.

PS: refactor can happen anytime. No need to be coupled.

Reply:
Yeah, I do want to begin having the same API with slightly different implementation for generating JSON vs XML. And agreed that it'll be a future TODO